scrape/scrape.go

```diff
@@ -1275,16 +1283,17 @@ const (
 	scrapeSeriesAddedMetricName  = "scrape_series_added" + "\xff"
 )
 
-func (sl *scrapeLoop) report(start time.Time, duration time.Duration, scraped, appended, seriesAdded int, scrapeErr error) (err error) {
+func (sl *scrapeLoop) report(app storage.Appender, start time.Time, duration time.Duration, scraped, appended, seriesAdded int, scrapeErr error) (err error) {
 	sl.scraper.Report(start, duration, scrapeErr)
 
 	ts := timestamp.FromTime(start)
 
 	var health float64
 	if scrapeErr == nil {
 		health = 1
+	} else {
+		app = sl.appender()
```

Comment:
Why is it in this `else`? Shouldn't we check for `app == nil`?

Reply:
When the scrape fails, the app is still not nil.
As mentioned in issue, use a new appender when the scraper fails.

Reply:
Spreading the logic across functions is fragile, I think this should all be in `run`

Reply:
> Spreading the logic across functions is fragile, I think this should all be in `run`
thanks .done.

![image](https://user-images.githubusercontent.com/9583245/78544249-6ebd1100-782c-11ea-8820-9c25e4345caa.png)
![image](https://user-images.githubusercontent.com/9583245/78544280-767cb580-782c-11ea-9f0a-7ba4b991f421.png)